### PR TITLE
Folder Factories

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
I always felt we were imprisoned by a boolean flag, hence this change.

Also, a bit of a longshot, but I am dropping the Folder Master business and calling things their real names, as I don't think Master has any better meaning than Read Only, actually the opposite, I think it causes for confusion as Master implies authoritative, which it isn't.

### CODE REVIEW

**Comment by calmh** [comment]
Haven't read the code yet but this sounds like a good idea. Probably negatory on the name change to read only though - that was the first name which literally no one understood. Take a look at the issue for sorting out the name, I think we settled on something like "send changes only" or so

**Comment by AudriusButkevicius** [comment]
I don't understand how people don't understand it tho.

We also have the help text at the bottom, explaining what it is and a link to help, is that not enough?

**Comment by AudriusButkevicius** [comment]
As an alternative we can call it `Read Only (Only send changes)` in the dropdown?

**Comment by Zillode** [comment]
Feels more like `Write Only` ;)

**Comment by lkwg82** [comment]
Without reading it yet. My first reaction was "finally someone else had the same idea". (I already started a  PR with the same intent.)

**Comment by AudriusButkevicius** [comment]
Reverted back to master.

File: lib/config/config.go [hunk 1]
```diff
@@ -23,6 +23,9 @@ const (
 	OldestHandledVersion = 10
 	CurrentVersion       = 13
 	MaxRescanIntervalS   = 365 * 24 * 60 * 60
+
+	FolderTypeReadWrite = "readwrite"
```
Comment (lkwg82) [inline lib/config/config.go]: These seem to fit better into `folder.go` (folderTypeXY to folder)

File: lib/config/config.go [hunk 2]
```diff
@@ -247,6 +250,15 @@ func convertV12V13(cfg *Configuration) {
 	cfg.Options.NATRenewalM = cfg.Options.DeprecatedUPnPRenewalM
 	cfg.Options.NATTimeoutS = cfg.Options.DeprecatedUPnPTimeoutS
 	cfg.Version = 13
+
+	for i, fcfg := range cfg.Folders {
+		if fcfg.DeprecatedReadOnly {
+			cfg.Folders[i].Type = FolderTypeReadOnly
```
Comment (lkwg82) [inline lib/config/config.go]: So Master-Folder (aka readOnly) become deprecated? Why?

Comment (AudriusButkevicius) [inline lib/config/config.go]: No, the boolean setting gets deprecated in favor of a string.

Comment (lkwg82) [inline lib/config/config.go]: ok

File: lib/config/folderconfiguration.go [hunk 3]
```diff
@@ -20,8 +20,8 @@ type FolderConfiguration struct {
 	ID                    string                      `xml:"id,attr" json:"id"`
 	Label                 string                      `xml:"label,attr" json:"label"`
 	RawPath               string                      `xml:"path,attr" json:"path"`
+	Type                  string                      `xml:"type,attr" json:"type"`
```
Comment (calmh) [inline lib/config/folderconfiguration.go]: Everything looks good to me, except the attribute type here. I don't think this should be a `string`, it should be a `FolderType` similar to `PullOrder`, which is an enum that knows how to serialize itself to/from text. Then the default value of zero equals `readwrite` and we don't run the risk of it suddenly having the value `"banana"` or `""` and panicing somewhere.

**Comment by calmh** [comment]
@st-review merge this, I'll fixup the type in next commit

**Comment by st-review** [comment]
:ok_hand: Merged as eabd2fc936be685a0b1207fde49dba77a012c2ff. Thanks, @AudriusButkevicius!

**Pull request closed without merging**